### PR TITLE
[mono-move] Check transaction argument values

### PR DESCRIPTION
**Pull request by vgao1996**
## Summary

Transaction arguments arrive as raw BCS bytes, and decoding them is not enough. AptosVM builds each argument by calling a framework constructor, which rejects values that decode perfectly well: a `String` must be valid UTF-8, an `Object<T>` must name an address that really holds one. Without those checks an entry function can receive a `String` of garbage bytes or an `Object` pointing at nothing.

This adds them, and admits public structs and enums as argument types while there.

- **Values are checked as they are decoded**, wherever they sit: in a vector, in an `Option`, or in a struct's fields. A bad value is refused before anything is built for it.
- **Which types are allowed is decided from the signature**, before any bytes are read. Public structs and enums qualify on AptosVM's terms: copyable, not a resource, public pack functions, fields that also qualify.
- **Scripts and entry functions are checked alike**, through one shared path, as AptosVM does. The value checks assume the type check already ran, so a payload that skipped it was not merely more permissive: a script parameter could name a type the argument bytes had no business building, `Coin` among them.
- **What the `Object` check reads is prepared when the function is loaded.** Confirming an object exists needs the layouts of `T` and of `ObjectCore`, neither of which the object's own fields mention. Lowering publishes both, so decoding never loads code.

## Deliberate differences from AptosVM

- We report `FAILED_TO_DESERIALIZE_ARGUMENT` where AptosVM aborts inside `0x1::string` or `0x1::object`.
- We refuse a type that *could* carry something inadmissible, even when the value does not: `Option<Private>` is refused even as `None`, and an enum is refused unless every variant is packable.
- AptosVM caps constructor calls per argument at 32. We cap nothing yet, and the existence reads are not gas-charged.

## Known gaps

Marked inline: no gas or bound on the existence reads, no caching of the per-type eligibility answer, no depth bound on its recursion, and untrusted structs lose the memcpy fast path.

## Test plan

Differential against `FakeExecutor`: malformed and valid `String` and `Object` arguments at top level and nested, an over-long `Option`, and public structs and enums including a generic, a partially packable enum, and the private, keyed, and non-copy refusals. Plus regression tests for six cases a review turned up, each verified to fail without its fix: `Object<u64>`, a struct with a function-typed field, a type whose fields share types (which hangs without memoization), the same `Object<T>` checked by two transactions in one block, and two scripts, one taking a `Coin` that would otherwise be minted straight from the argument bytes and one taking an `Object` of a primitive.

Executor, runtime, core, loader, output, specializer, and the differential tests pass; clippy and fmt clean.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **High Risk**
> Changes security-critical user-input decoding and admission rules (strings, objects, public structs) and performs storage reads during argument construction without noted gas bounds.
> 
> **Overview**
> **Closes the gap between raw BCS decoding and AptosVM-style argument safety** for entry functions and scripts: signature checks and value validation now run before execution, with scripts getting the same parameter-type rules as entry functions (e.g. blocking forged `Coin` arguments).
> 
> **Signature validation** moves into `args.rs` and is shared by entry functions and scripts. It enforces leading signers, whitelisted framework types, and **public copyable structs/enums** that expose public `pack` functions, with memoized type walks and stricter `Object<T>` rules (`T` must be a nominal resource type). Argument placement uses **`arg_bcs_untrusted`** instead of plain BCS decode.
> 
> **Runtime/core** add untrusted deserialization hooks that validate **`String` UTF-8** and **`Object<T>`** via storage existence checks for `ObjectCore` and `T`. `ValueLayout` now records its **`ty`** so checks run during decode (including nested vectors/options/structs); lowering publishes `Object` resource layouts ahead of checks. New executor errors map to **`FAILED_TO_DESERIALIZE_ARGUMENT`** (and related invalid-signature codes).
> 
> **Tests** add broad differential e2e coverage against v1 for strings, objects, options, public types, script forgery, and edge cases (shared-field chains, repeated `Object<T>` in a block).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 6dc050b09e3c89b19c81218886f4c305fae704fa. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: third_party/move/mono-move/aptos-transaction-executor/src/user_txn/entry_func.rs [hunk 1]
```diff
@@ -80,41 +91,138 @@ fn is_allowed_arg_type(ty: InternedType) -> bool {
...
+/// Whether a struct or enum outside the framework whitelist can be allowed as
+/// a transaction argument.
+/// - Its definition must declare `copy` and not `key`.
+/// - It must have public pack function: `pack$S` (if struct), or pack functions for
+//    every variant (if enum).
+/// - Its field types must all be allowed.
+//
+// TODO(perf): the answer depends only on the type, so cache it instead of
+// walking the defining module's functions and fields on every transaction.
+fn is_allowed_nominal_type<'a>(
```
Comment (vgao1996) [inline third_party/move/mono-move/aptos-transaction-executor/src/user_txn/entry_func.rs]: @georgemitenkov @vineethk Thoughts on performing this check here v.s. caching it in the global context.

File: third_party/move/mono-move/runtime/src/error.rs [hunk 2]
```diff
@@ -114,6 +116,25 @@ pub enum RuntimeError {
...
+    /// An `Object` argument names an address holding no object.
+    #[error("argument check: no object at the `Object` argument's address")]
+    ObjectArgumentDoesNotExist,
+
+    /// An `Object<T>` argument names an object holding no `T`.
+    #[error("argument check: the `Object` argument's address holds no resource of its type")]
+    ObjectArgumentLacksResource,
+
+    /// A storage read while checking an `Object` argument failed.
+    #[error("argument check: {0}")]
```
Comment (georgemitenkov) [inline third_party/move/mono-move/runtime/src/error.rs]: This should be propagated as storage failure? Also instead of "argument check: {0}" maybe "storage read failure during `Object` argument check: {0}"

File: third_party/move/mono-move/specializer/src/lower/context.rs
Comment (georgemitenkov) [inline third_party/move/mono-move/specializer/src/lower/context.rs:1171]: We should make sure we discover args/returns of entry function calling into interpreter in this pass.

File: third_party/move/mono-move/specializer/src/lower/context.rs [hunk 3]
```diff
@@ -1097,6 +1097,26 @@ pub fn try_discover_types_for_lowering_in_module(
     Ok(descriptors)
 }
 
+/// Publishes a resource type's layout and struct descriptor, so it can be
+/// read from storage outside lowered code.
+pub fn try_publish_resource_type(
```
Comment (georgemitenkov) [inline third_party/move/mono-move/specializer/src/lower/context.rs]: I think this is wrong to make it a separate in BCS hook. We should at load time get all the info we need.

File: third_party/move/mono-move/loader/src/loader.rs [hunk 4]
```diff
@@ -259,6 +261,48 @@ impl<'guard, 'ctx> Loader<'guard, 'ctx> {
         Ok(module.set_instantiated_function(func_name, ty_args, function, function_ms))
     }
 
+    pub fn publish_resource_type(
```
Comment (georgemitenkov) [inline third_party/move/mono-move/loader/src/loader.rs]: Why do we need this? When function is loaded all types are loaded so if something is missing there it should be added to lowering context (which might not load all the things we need right now). So invariant is that ALL entry function arguments have their nested types/layout known.

Doing this at runtime complicates semantics, metering, etc.

File: third_party/move/mono-move/core/src/value_layout.rs [hunk 5]
```diff
@@ -368,26 +383,29 @@ impl ValueLayout {
...
             align,
             fixed_bcs_size: Some(size),
             flags: LayoutFlags::NO_POINTERS_NO_PADDING | LayoutFlags::ALL_BYTE_PATTERNS_VALID,
             kind: LayoutKind::UnsignedInt,
+            ty: Some(ty),
         }
     }
 
-    fn signed_int(size: u32, align: u32) -> Self {
+    fn signed_int(size: u32, align: u32, ty: InternedType) -> Self {
```
Comment (georgemitenkov) [inline third_party/move/mono-move/core/src/value_layout.rs]: passing the type here somewhat kills the purpose of these functions, because we can do `signed_int(ty.size(), ty.align(), ty)` and there is no longer an invariant that those match? Consider changing constructors then.

File: third_party/move/mono-move/core/src/value_layout.rs [hunk 6]
```diff
@@ -328,6 +341,7 @@ impl ValueLayout {
                 elem_id,
                 descriptor_id,
             },
+            ty: None,
```
Comment (georgemitenkov) [inline third_party/move/mono-move/core/src/value_layout.rs]: I think we should always set this type to some and elsewhere at construction, otherwise caller can do whatever they want and there is no invariant type and element descriptor match?

File: third_party/move/mono-move/core/src/value_layout.rs [hunk 7]
```diff
@@ -72,13 +76,18 @@ pub struct ValueLayout {
...
+    /// them).
     pub fixed_bcs_size: Option<u32>,
     /// Flags with extra information about this layout / type.
     pub flags: LayoutFlags,
     /// Describes layout's shape.
     pub kind: LayoutKind,
+    /// The type this layout describes.
+    ///
+    /// Set to [`None`] for
+    /// - references and function values, as they use shared layouts
```
Comment (georgemitenkov) [inline third_party/move/mono-move/core/src/value_layout.rs]: references have no layout?

File: third_party/move/mono-move/global-context/src/context.rs [hunk 8]
```diff
@@ -713,10 +713,11 @@ impl<'ctx> ExecutionGuard<'ctx> {
 
     /// Publishes the layout for the given type and returns its assigned
     /// [`LayoutId`].
-    pub fn publish_layout(&self, ty: InternedType, layout: ValueLayout) -> LayoutId {
+    pub fn publish_layout(&self, ty: InternedType, mut layout: ValueLayout) -> LayoutId {
```
Comment (georgemitenkov) [inline third_party/move/mono-move/global-context/src/context.rs]: Since we store the type, I wonder if we can just publish layout by `layout.ty` then so that all APIs use same type used at construction?

Also, should we reverse type info into a separate table? Because we then increase layout size of our struct/enum, and type info is not used often (only for a few args) so ideally we decouple that so we are more cache/memory friendly

File: third_party/move/mono-move/aptos-transaction-executor/src/user_txn/entry_func.rs (hunk 1 above)
Comment (georgemitenkov) [reply third_party/move/mono-move/aptos-transaction-executor/src/user_txn/entry_func.rs]: We can decide that when we get back to code gen solution?

Comment (georgemitenkov) [reply third_party/move/mono-move/aptos-transaction-executor/src/user_txn/entry_func.rs]: I think here is fine for now, we also have tags, state keys to cache so better do these together.

File: third_party/move/mono-move/global-context/src/context.rs (hunk 8 above)
Comment (vgao1996) [reply third_party/move/mono-move/global-context/src/context.rs]: > Since we store the type, I wonder if we can just publish layout by layout.ty then so that all APIs use same type used at construction?

Yes.

> reverse type info into a separate table

I'll leave a TODO for revisiting in the future, but for now there is not much benefit storing this in a side table, as we store the type info for all layouts that are eligible, which covers most layouts. In other words, the reverse mapping is dense.

File: third_party/move/mono-move/loader/src/loader.rs (hunk 4 above)
Comment (vgao1996) [reply third_party/move/mono-move/loader/src/loader.rs]: Got rid of this and moved discovery entirely into the specializer: for every `Object<T>` we discover `ObjectCore` and `T` as well. There are some minor inefficiencies here, but not something to worry about for now.

File: third_party/move/mono-move/core/src/value_layout.rs (hunk 7 above)
Comment (vgao1996) [reply third_party/move/mono-move/core/src/value_layout.rs]: All reference type share the same layout. There is no single type the reference layout can map back to. Revised the doc comment to state this more clearly.

**Comment by vgao1996** [comment]
@georgemitenkov addressed your current comments + fixed a major bug (scripts not performing the same arg validation)
